lib/predictions/predictions.ex

```diff
@@ -38,77 +26,98 @@ defmodule Predictions.Predictions do
      end), vehicles_running_revenue_trips}
   end
 
-  @spec transform_stop_time_updates(map()) :: [
-          {map(), String.t(), String.t(), integer(), String.t(), boolean(), String.t() | nil}
-        ]
-  defp transform_stop_time_updates(trip_update) do
-    last_stop_id =
-      Enum.max_by(trip_update["stop_time_update"], fn update ->
-        if update["arrival"], do: update["arrival"]["time"], else: 0
-      end)
-      |> Map.get("stop_id")
-
-    vehicle_id = get_in(trip_update, ["vehicle", "id"])
-
-    Enum.map(
-      trip_update["stop_time_update"],
-      &{&1, last_stop_id, trip_update["trip"]["route_id"], trip_update["trip"]["direction_id"],
-       trip_update["trip"]["trip_id"], trip_update["trip"]["revenue"], vehicle_id}
-    )
+  defp valid_trip_update?(trip_update) do
+    relevant_rail_route?(trip_update["trip"]["route_id"]) and
+      trip_update["trip"]["schedule_relationship"] != "CANCELED"
   end
 
-  @spec prediction_from_update(
-          {map(), String.t(), String.t(), integer(), Predictions.Prediction.trip_id(), boolean(),
-           String.t() | nil},
-          DateTime.t()
-        ) :: Prediction.t()
-  defp prediction_from_update(
-         {stop_time_update, last_stop_id, route_id, direction_id, trip_id, revenue_trip?,
-          vehicle_id},
-         current_time
-       ) do
+  @spec trip_update_to_predictions(map(), DateTime.t()) :: [Prediction.t()]
+  defp trip_update_to_predictions(trip_update, current_time) do
     current_time_seconds = DateTime.to_unix(current_time)
+    destination_stop_id = get_destination_stop_id(trip_update)
+    vehicle_id = trip_update["vehicle"]["id"]
+    vehicle_location = Engine.Locations.for_vehicle(vehicle_id)
+    route_id = trip_update["trip"]["route_id"]
+    direction_id = trip_update["trip"]["direction_id"]
+    trip_id = trip_update["trip"]["trip_id"]
+    revenue_trip? = trip_update["trip"]["revenue"]
```

Comment:
nit: We could inline some of these below to shorten the boilerplate a bit. Alternatively, we could destructure the function param directly, as long as these fields will always be present.